File: src/test/java/org/cip4/xjdf/json/openapi/ComplexTypeTest.java
package org.cip4.xjdf.json.openapi;

import org.cip4.xjdf.json.openapi.model.Model;
import org.cip4.xjdf.json.openapi.model.Schema;
import org.cip4.xjdf.json.openapi.model.Schemas;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.xml.sax.InputSource;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;
import java.io.InputStream;
import java.util.Map;

public class ComplexTypeTest {
    private final TypeTranslator nameTranslator = new TypeTranslator("#/defs");

    @Test
    public void emptyComplexTypeIsHandled() throws Exception {
        Document doc;
        try (InputStream is = getClass().getResourceAsStream("/abstract-element.xml")) {
            doc = readXml(is);
        }

        XPathFactory xpFactory = XPathFactory.newInstance();
        javax.xml.xpath.XPath xPath = xpFactory.newXPath();
        xPath.setNamespaceContext(new XsdNamespaceContext());

        ComplexType tle = new ComplexType(
            new Context(
                xPath,
                nameTranslator,
                (Node) xPath.evaluate("/xs:schema/xs:complexType[@name='AbstractType']", doc, XPathConstants.NODE)
            )
        );
        Model model = tle.getModel();
        Assertions.assertEquals(
            new Model(
                "AbstractType",
                new Schema().type("object")
            ),
            model
        );
    }

    @Test
    public void inheritingTypeReferencesParent() throws Exception {
        Document doc;
        try (InputStream is = getClass().getResourceAsStream("/abstract-element.xml")) {
            doc = readXml(is);
        }

        XPathFactory xpFactory = XPathFactory.newInstance();
        javax.xml.xpath.XPath xPath = xpFactory.newXPath();
        xPath.setNamespaceContext(new XsdNamespaceContext());

        ComplexType tle = new ComplexType(
            new Context(
                xPath,
                nameTranslator,
                (Node) xPath.evaluate("/xs:schema/xs:complexType[@name='Impl1Type']", doc, XPathConstants.NODE)
            )
        );
        Model model = tle.getModel();
        Assertions.assertEquals(
            new Model(
                "Impl1Type",
                new Schema()
                    .allOf(
                        nameTranslator.translate("AbstractType"),
                        new Schema()
                            .type("object")
                            .properties(new Schemas(Map.of(
                                "Foo", new Schema().type("string")
                            )))
                    )
            ),
            model
        );
    }

    private Document readXml(InputStream sourceXsd) throws Exception {
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        dbFactory.setNamespaceAware(true);
        javax.xml.parsers.DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        InputSource xmlInput = new InputSource(sourceXsd);

        return dBuilder.parse(xmlInput);
    }
}

File: src/test/kotlin/org/cip4/xjdf/openapi/ComplexTypeTest.kt
/*
 * The CIP4 Software License, Version 1.0
 *
 * Copyright (c) 2001-2021 The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress (CIP4). All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * 3. The end-user documentation included with the redistribution, if any, must
 * include the following acknowledgment: &quot;This product includes software developed
 * by the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress (www.cip4.org)&quot; Alternately, this acknowledgment
 * may appear in the software itself, if and wherever such third-party
 * acknowledgments normally appear.
 *
 * 4. The names &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress&quot; must not be used to endorse or
 * promote products derived from this software without prior written permission.
 * For written permission, please contact info@cip4.org
 *
 * 5. Products derived from this software may not be called &quot;CIP4&quot;, nor may &quot;CIP4&quot;
 * appear in their name, without prior written permission of the CIP4 organization
 *
 * &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress&quot; are trademarks of The International Cooperation
 * for the Integration of Processes in Prepress, Press and Postpress (CIP4).  All
 * other names and brands are the property of their respective owners.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. CIP4 AND ITS CONTRIBUTORS MAKE
 * NO REPRESENTATION THAT THE USE OF THIS SOFTWARE OR DATA WILL NOT INFRINGE ANY
 * PATENTS, COPYRIGHTS, TRADEMARKS, OR OTHER RIGHTS. IN NO EVENT SHALL CIP4 OR ITS
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many individuals on
 * behalf of the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress and was originally based on software some of which
 * contributed by and copyrighted by Adobe Systems and Man Roland. copyright (c)
 * 2000-2001, Adobe Systems copyright (c) 2001-2001, Man Roland
 *
 * For more information on The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress , please see &lt;http://www.cip4.org/&gt;.
 */

package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.NamedSchema
import org.cip4.xjdf.openapi.model.Schema
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.xml.sax.InputSource
import java.io.InputStream
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathFactory

internal class ComplexTypeTest {

    @Test
    internal fun `empty complex type is handled`() {
        val doc = readXml(this::class.java.getResource("/abstract-element.xml").openStream())

        val xpFactory = XPathFactory.newInstance()
        val xPath = xpFactory.newXPath()
        xPath.namespaceContext = XsdNamespaceContext()

        val tle = ComplexType(
            Context(
                xPath,
                TypeTranslator(),
                xPath.evaluate("/xs:schema/xs:complexType[@name='AbstractType']", doc, XPathConstants.NODE) as Node
            )
        )
        val model = tle.getModel()
        assertEquals(
            NamedSchema(
                "AbstractType",
                Schema(
                    type = "object"
                )
            ),
            model
        )
    }

    @Test
    internal fun `inheriting type references parent`() {
        val doc = readXml(this::class.java.getResource("/abstract-element.xml").openStream())

        val xpFactory = XPathFactory.newInstance()
        val xPath = xpFactory.newXPath()
        xPath.namespaceContext = XsdNamespaceContext()

        val tle = ComplexType(
            Context(
                xPath,
                TypeTranslator(),
                xPath.evaluate("/xs:schema/xs:complexType[@name='Impl1Type']", doc, XPathConstants.NODE) as Node
            )
        )
        val model = tle.getModel()
        assertEquals(
            NamedSchema(
                "Impl1Type",
                Schema(
                    allOf = listOf(
                        Schema.ref("AbstractType"),
                        Schema(
                            type = "object",
                            properties = mapOf(
                                Pair("Foo", Schema(type = "string"))
                            )
                        )
                    )
                )
            ),
            model
        )
    }

    private fun readXml(sourceXsd: InputStream): Document {
        val dbFactory = DocumentBuilderFactory.newInstance()
        dbFactory.isNamespaceAware = true
        val dBuilder = dbFactory.newDocumentBuilder()
        val xmlInput = InputSource(sourceXsd)

        return dBuilder.parse(xmlInput)
    }

}